pandas/core/arrays/categorical.py

```diff
@@ -400,20 +400,38 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
             If copy is set to False and dtype is categorical, the original
             object is returned.
         """
-        if is_categorical_dtype(dtype):
+        if self.dtype is dtype:
+            result = self.copy() if copy else self
+
+        elif is_categorical_dtype(dtype):
             dtype = cast(Union[str, CategoricalDtype], dtype)
 
             # GH 10696/18593
             dtype = self.dtype.update_dtype(dtype)
             self = self.copy() if copy else self
-            if dtype == self.dtype:
-                return self
-            return self._set_dtype(dtype)
-        if is_extension_array_dtype(dtype):
-            return array(self, dtype=dtype, copy=copy)
-        if is_integer_dtype(dtype) and self.isna().any():
+            result = self._set_dtype(dtype)
+
+        elif is_extension_array_dtype(dtype):
+            result = array(self, dtype=dtype, copy=copy)
+
+        elif is_integer_dtype(dtype) and self.isna().any():
             raise ValueError("Cannot convert float NaN to integer")
-        return np.array(self, dtype=dtype, copy=copy)
+
+        elif len(self.codes) == 0 or len(self.categories) == 0:
+            result = np.array(self, dtype=dtype, copy=copy)
+
+        else:
+            # GH8628 (PERF): astype category codes instead of astyping array
+            try:
+                astyped_categories = self.categories.astype(dtype=dtype, copy=copy)
+            except (TypeError, ValueError):
+                raise ValueError(
+                    f"Cannot cast {self.categories.dtype} dtype to {dtype}"
+                )
+            astyped_categories = extract_array(astyped_categories, extract_numpy=True)
+            result = np.array(take_1d(astyped_categories, self._codes), copy=copy)
```

Comment:
pretty sure you are copying twice here as the take_1d copies.

Reply:
of if we could have an EA returned, then do `np.asarray`

Reply:
EA case is handled above so I removed here

But probably it's possible to consolidate EA/non-EA and get the same speed up for EAs (will look in follow-on)

Reply:
> But probably it's possible to consolidate EA/non-EA and get the same speed up for EAs (will look in follow-on)

Added a TODO for this